src/main/scala/abalone/variant/Variant.scala

```diff
@@ -193,72 +192,86 @@ abstract class Variant private[variant] (
           case (pos, neighbour) =>
             generateMovesForNeighbours(pos, neighbour)
         }).view.toList
-      }
+    }
   }
 
-  // @TODO: would be interesting to pass the direction so we can easily differenciate a side move from a line move (orig meets dest in case of line move)
-  def boardAfter(situation: Situation, orig: Pos, dest: Pos): Board = {
+  def boardAfter(situation: Situation, orig: Pos, dest: Pos, directions: List[String] = List()): Board = {
     // 1. move pieces and get the score
-    val (pieces, capture)   = piecesAfterAction(situation.board.pieces, situation.player, orig, dest)
-
-    // 2. update the board
-    situation.board.copy(
-      pieces = pieces,
-      history = History(
-        situation.history.currentTurn,
-        situation.history.currentTurn, // @TODO: fix this, understand how to create the new Uci with the move being played
-        situation.history.positionHashes,
-        if (capture) situation.history.score.add(situation.player) else situation.history.score,
-        situation.history.halfMoveClock + 1
+    val (pieces, capture)   = piecesAfterAction(situation.board.pieces, orig, dest, directions)
+    val score = Score(situation.history.score.p1, situation.history.score.p2)
+
+    val boardAfter = situation.board.copy(pieces = pieces)
+    boardAfter.withHistory(
+      situation.history.copy(
+        // @TODO: lastTurn handled on Move.finalizeAfter to keep consistency with other gamelogics ?
+        score = if (capture) score.add(situation.player) else score,
+        halfMoveClock = situation.board.history.halfMoveClock + 1 // situation.player.fold(0, 1)
       )
     )
   }
 
-  // @TODO: adapt
+  // @TODO: rewrite this code properly as it's currently full of get()
   // will take care of moving the marbles and determine if a capture has been made
-  def piecesAfterAction(pieces: PieceMap, player: Player, @nowarn orig: Pos, dest: Pos): (PieceMap, Boolean) = {
+  private def piecesAfterAction(pieces: PieceMap, orig: Pos, dest: Pos, directions: List[String]): (PieceMap, Boolean) = {
     var capture = false
+    var updatedPieceMap: PieceMap = Map()
 
     /*
-
-        * * * * *
-       * a A B C 1
-      * * * * * 2 *
-      How to move pieces based on orig and dest :
-        1. identify the type of move :
-          - push : dest contains a marble (orig=C, dest=a)
-          - line move : orig meets dest thanks to the direction we passed as extra parameter
-          - side move : NOT a push or a side move
+      How to move pieces based on orig, dest, and direction :
+        1. identify the type of move
+          - line move : no direction to apply (a line move is just 1 marble moving)
+          - push : 1 direction, and dest contains a marble
+          - side move : 2 direction:
```

Comment:
is the direction not the same for the side move pieces, so why not just pass in one direction? Given you don't want a direction for line move, you can now have Option[String], and then match on this value for your cases, instead of if statements. A direction must match the direction of orig dest if its a push, otherwise its a side move?

Reply:
because I want to be able to find back all the pieces quickly, then apply the direction of the side move to all these

Reply:
ah the directions is storing more info as it also stores if more pieces are moving... great thanks.